dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala

```diff
@@ -21,18 +21,71 @@ import com.intel.analytics.bigdl.Module
 import com.intel.analytics.bigdl.tensor.Tensor
 import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
 import com.intel.analytics.bigdl.utils.T
+import com.intel.analytics.bigdl.utils.ThreadPool._
+import org.apache.log4j.Logger
 
 import scala.reflect.ClassTag
 
+/**
+ * @param hiddenSize
+ * @param bpttTruncate
+ * @param timeDim the time dimension of the input Tensor
+ * @param ev
+ * @tparam T
+ */
+
 class Recurrent[T : ClassTag] (
   hiddenSize: Int = 3,
-  bpttTruncate: Int = 2)
+  bpttTruncate: Int = 2,
+  timeDim: Int = 2)
   (implicit ev: TensorNumeric[T]) extends Container[Tensor[T], Tensor[T], T] {
 
+  /**
+   * if transpose is set to be true, the Recurrent layer will transpose the
+   * input tensor. Thus, the legacyMode must be set to be true to select the
+   * correct dimension for recurrence
+   *
+   * if transpose is set to be false, there are two probable scenarios.
+   *
+   * For the first scenario, the input tensor might be transpose before recurrent
+   * layer. Thus, the legacyMode should be set to be true so as to select the
+   * correct dimension for recurrence.
+   *
+   * For the second one, the input tensor has not been transpose, and user can
+   * define legacyMode to be false to select correct time dimension for recurrence.
+   * The drawback of this mode is that the storage in the output Tensor will be
+   * discontiguous, which will slow the processing speed.
+   */
+
+  private var legacyMode: Boolean = false
   val hidden = Tensor[T]()
   var module: Module[T] = _
   var transform: Module[T] = _
-  var (batchSize, times) = (0, 0)
+  var (batchSize, times, nDim) = (0, 0, 0)
+
+  /**
+   * batchDim should be either 1 or 2 according to timeDim
+   * hiddenSelect is always equal to 1
+   * outputSelect is always equal to timeDim
+   */
+
+  val batchDim = 3 - timeDim
```

Comment:
The input should be a 3D tensor, so it is not possible for the timeDim set to 4.
I have added a requirement for this argument.